### PR TITLE
Conversion webhook

### PR DESCRIPTION
**Pull request by bshifter**
Conversion webhook as an opt-in feature and v1alpha1 versions (go back in time) identical to v1beta1 to be able to test and roll out the whole procedure through resourcebuilders. This change will not affect default installations other than the chart will create v1alpha1 CRDs as well, although those shouldn't affect existing clients by any means.

### CODE REVIEW

**Review by pepov — APPROVED** [review]

File: main.go [hunk 1]
```diff
@@ -105,8 +107,16 @@ func main() {
 		setupLog.Error(err, "unable to create controller", "controller", "Logging")
 		os.Exit(1)
 	}
+
 	// +kubebuilder:scaffold:builder
```
Comment (siliconbrain) [inline main.go:111]:
```suggestion
```
I think there should be only one `+kubebuilder:scaffold:builder` marker.

File: pkg/sdk/resourcebuilder/component.go [hunk 2]
```diff
@@ -31,14 +31,17 @@ import (
...
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	serializer "k8s.io/apimachinery/pkg/runtime/serializer/json"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
 )
 
 const (
-	Image            = "ghcr.io/banzaicloud/logging-operator:3.10.0"
+	Image            = "ghcr.io/banzaicloud/logging-operator:3.9.5-dev"
```
Comment (siliconbrain) [inline pkg/sdk/resourcebuilder/component.go:44]: Is this the image version we're going with?

File: pkg/sdk/resourcebuilder/component.go [hunk 3]
```diff
@@ -132,6 +141,15 @@ func CRD(config *ComponentConfig, group string, kind string) (runtime.Object, re
 		return nil, nil, errors.WrapIff(err, "failed to read %s crd", kind)
 	}
 
+	// apply modifiers
+	for _, modifier := range modifiers {
+		CRDmodified, err := modifier(crd)
```
Comment (siliconbrain) [inline pkg/sdk/resourcebuilder/component.go:146]:
```suggestion
		crdModified, err := modifier(crd)
```
?

File: pkg/sdk/resourcebuilder/component.go [hunk 4]
```diff
@@ -299,3 +335,78 @@ func (c *ComponentConfig) labelSelector(parent reconciler.ResourceOwner) map[str
...
+				Name:       "conversion-webhook",
+				Port:       DefaultSecurePort,
+				TargetPort: intstr.IntOrString{IntVal: DefaultWebhookPort},
+				Protocol:   corev1.ProtocolTCP,
+			},
+		},
+		Selector: config.labelSelector(parent),
+		Type:     corev1.ServiceTypeClusterIP,
+	}
+	svc.Spec = spec
```
Comment (siliconbrain) [inline pkg/sdk/resourcebuilder/component.go:363]:
```suggestion
	svc.Spec = corev1.ServiceSpec{
		Ports: []corev1.ServicePort{
			{
				Name:       "conversion-webhook",
				Port:       DefaultSecurePort,
				TargetPort: intstr.IntOrString{IntVal: DefaultWebhookPort},
				Protocol:   corev1.ProtocolTCP,
			},
		},
		Selector: config.labelSelector(parent),
		Type:     corev1.ServiceTypeClusterIP,
	}
```

File: pkg/sdk/resourcebuilder/webhooks.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		return crd, nil
+	}
+}
+
+func ModifierCAInjectAnnotation(certName types.NamespacedName) CRDModifier {
+	return func(crd *v1.CustomResourceDefinition) (*v1.CustomResourceDefinition, error) {
+		if crd.Annotations == nil {
+			crd.Annotations = make(map[string]string)
+		}
+		crd.Annotations[CertManagerInjectCAFromAnnotationKey] = fmt.Sprintf("%s/%s", certName.Namespace, certName.Name)
```
Comment (siliconbrain) [inline pkg/sdk/resourcebuilder/webhooks.go:68]:
```suggestion
		crd.Annotations[CertManagerInjectCAFromAnnotationKey] = certName.String()
```

**Review by siliconbrain — APPROVED** [review]

File: pkg/sdk/resourcebuilder/component.go (hunk 2 above)
Comment (pepov) [reply pkg/sdk/resourcebuilder/component.go:44]: No, I think we will need an even newer tagged image for this, possibly 3.11.0

**Pull request merged**